### PR TITLE
Feature implementation: release list charts

### PR DESCRIPTION
**Pull request by nicholasSUSE**
#### Issue:

https://github.com/rancher/ecm-distro-tools/issues/445

#### Solution:

After implementing: https://github.com/rancher/charts-build-scripts/pull/144

This is the integration between: 
- `charts-build-scripts` command `lifecycle-status`  
- `ecm-distro-tools` command `release list charts`

#### Result example:

The logs printed by `lifecycle-status` are printed to the terminal and the saved log files path is also printed in the terminal:
```
.
.
.

time=2024-07-26T18:11:33-03:00 level=info msg=Assets to be RELEASED

time=2024-07-26T18:11:33-03:00 level=info msg=rancher-istio

__________________________________________________________________
time=2024-07-26T18:11:33-03:00 level=info msg=Assets to be FORWARD-PORTED

time=2024-07-26T18:11:33-03:00 level=info msg=rancher-istio
time=2024-07-26T18:11:33-03:00 level=info msg=saving app state to state.json file
time=2024-07-26T18:11:33-03:00 level=info msg=saved state file successfully

generated log files for inspection at: 
/home/nick/WORK/SOFTWARE/go/src/github.com/release-work/charts/logs/
```

### CODE REVIEW

File: cmd/release/cmd/list.go [hunk 1]
```diff
@@ -40,8 +41,38 @@ var rancherListRCDepsSubCmd = &cobra.Command{
 	},
 }
 
+var chartsListSubCmd = &cobra.Command{
+	Use:   "charts [branch] [charts](optional)",
+	Short: "List Charts assets versions state for release process",
+	RunE: func(cmd *cobra.Command, args []string) error {
+
+		var branch, chart string = "", ""
```
Comment (briandowns) [inline cmd/release/cmd/list.go:49]: Let's take advantage of the "0 value" and update this to be

```go

var branch string
var chart string
```

We don't need to explicitly set the new string to ""

File: cmd/release/cmd/list.go [hunk 2]
```diff
@@ -40,8 +41,38 @@ var rancherListRCDepsSubCmd = &cobra.Command{
...
+		if len(args) > 1 {
+			chart = args[1]
+		}
+
+		config := rootConfig.Charts
+		if config.Workspace == "" || config.ChartsForkURL == "" {
+			return errors.New("verify your config file, chart configuration not implemented correctly, you must insert workspace path and your forked repo url")
+		}
+
+		err := charts.ListLifecycleStatus(context.Background(), config, branch, chart)
```
Comment (briandowns) [inline cmd/release/cmd/list.go:65]: You can return here and remove the code underneath.

File: cmd/release/cmd/list.go [hunk 3]
```diff
@@ -40,8 +41,38 @@ var rancherListRCDepsSubCmd = &cobra.Command{
...
+
+		if len(args) < 1 {
+			return errors.New("expected at least one argument: [branch]")
+		}
+		branch = args[0]
+
+		if len(args) > 1 {
+			chart = args[1]
+		}
+
```
Comment (briandowns) [inline cmd/release/cmd/list.go:59]: Should we have any validations on input here? Ie validate a valid branch or chart has been provided?

File: release/charts/charts.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
+package charts
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	ecmConfig "github.com/rancher/ecm-distro-tools/cmd/release/config"
```
Comment (briandowns) [inline release/charts/charts.go:10]: Does this need the import alias?

File: release/charts/charts.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+)
+
+// ListLifecycleStatus prints the lifecycle status of the charts
+func ListLifecycleStatus(ctx context.Context, c *ecmConfig.ChartsRelease, branch, chart string) error {
+
```
Comment (briandowns) [inline release/charts/charts.go:15]: Remove this new line.

File: release/charts/charts.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	var branchArg, chartArg string = "", ""
```
Comment (briandowns) [inline release/charts/charts.go:16]: Same comment as above.

File: release/charts/charts.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+	branchArg = "--branch-version=" + branch
+	if chart != "" {
+		chartArg = "--chart=" + chart
+	}
+
+	err := executeChartsBuildScripts(c.Workspace, "lifecycle-status", branchArg, chartArg)
+	if err != nil {
```
Comment (briandowns) [inline release/charts/charts.go:24]: Move the if check up and do it inline scope locking the err value.

File: release/charts/charts.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+	fmt.Printf("generated log files for inspection at: \n%s\n", c.Workspace+"/logs/")
+	return nil
+}
+
+func executeChartsBuildScripts(chartsRepoPath string, args ...string) error {
+	// save current working dir
+	ecmWorkDir, _ := os.Getwd()
+
+	// change working dir to the charts repo
```
Comment (briandowns) [inline release/charts/charts.go:39]: What happens if there's an error?

File: release/charts/charts.go [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+func executeChartsBuildScripts(chartsRepoPath string, args ...string) error {
+	// save current working dir
+	ecmWorkDir, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	// change working dir to the charts repo
+	err = os.Chdir(chartsRepoPath)
```
Comment (briandowns) [inline release/charts/charts.go:40]: Move the if check up a line and scope lock the err value. Please do this for all calls that return a single err value.

File: release/charts/charts.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	if err != nil {
+		return err
+	}
+
+	bin := strings.Join([]string{chartsRepoPath, "bin", "charts-build-scripts"}, string(os.PathSeparator))
+
+	cmd := exec.Command(bin, args...)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
```
Comment (briandowns) [inline release/charts/charts.go:49]: What is the contents of err if an error could occur?

File: release/charts/charts.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+		return fmt.Errorf("failed to execute charts-build-scripts: %w, output: %s", err, output)
+	}
+	fmt.Printf("charts-build-scripts output: %s\n", output)
+
+	// Change back working dir for the caller
```
Comment (briandowns) [inline release/charts/charts.go:54]: You can return on this line.

File: release/charts/charts.go [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+	var branchArg, chartArg string = "", ""
+
+	branchArg = "--branch-version=" + branch
+	if chart != "" {
+		chartArg = "--chart=" + chart
+	}
+
+	err := executeChartsBuildScripts(c.Workspace, "lifecycle-status", branchArg, chartArg)
+	if err != nil {
```
Comment (briandowns) [inline release/charts/charts.go:24]: Let's rename this to: runChartsBuild. The previous name leaks implementation detail that could change in the future. This new name still conveys the idea and allows us to change from scripts to programs and exec model.

**Review by briandowns — CHANGES_REQUESTED** [review]

File: cmd/release/cmd/list.go (hunk 3 above)
Comment (nicholasSUSE) [reply cmd/release/cmd/list.go:59]: So, you got me thinking here, if a wrong chart is provided, this is the response:
```
Error: failed to execute charts-build-scripts: exit status 1, output: time="2024-07-26T18:33:23-03:00" level=info msg="Initializing dependencies for lifecycle-status"
time=2024-07-26T18:33:23-03:00 level=fatal msg=encountered error while initializing dependencies: no assets found in the repository

error: failed to execute charts-build-scripts: exit status 1, output: time="2024-07-26T18:33:23-03:00" level=info msg="Initializing dependencies for lifecycle-status"
time=2024-07-26T18:33:23-03:00 level=fatal msg=encountered error while initializing dependencies: no assets found in the repository
```
I agree this is no good, so

I have two options here: 
- either I change the response at `charts-build-scripts` when a wrong chart is provided or,
- I implement a scan of ecm-distro-tools release list charts at the `charts-repo`

File: release/charts/charts.go (hunk 4 above)
Comment (nicholasSUSE) [reply release/charts/charts.go:10]: To be honest I copied it from: 
`release/k3s/k3s.go` but I can remove it.
I just wanted to follow the pattern that was already there from the example I caught.

File: release/charts/charts.go (hunk 8 above)
Comment (nicholasSUSE) [reply release/charts/charts.go:37]: @briandowns 
There is no way to continue the execution if there is an error when discovering the current dir and changing dirs.

File: release/charts/charts.go (hunk 10 above)
Comment (nicholasSUSE) [reply release/charts/charts.go:46]: @briandowns: One small example:
```
Error: failed to execute charts-build-scripts: exit status 1, output: time="2024-07-26T18:33:23-03:00" level=info msg="Initializing dependencies for lifecycle-status"
time=2024-07-26T18:33:23-03:00 level=fatal msg=encountered error while initializing dependencies: no assets found in the repository

error: failed to execute charts-build-scripts: exit status 1, output: time="2024-07-26T18:33:23-03:00" level=info msg="Initializing dependencies for lifecycle-status"
time=2024-07-26T18:33:23-03:00 level=fatal msg=encountered error while initializing dependencies: no assets found in the repository
```

File: cmd/release/cmd/list.go (hunk 3 above)
Comment (briandowns) [reply cmd/release/cmd/list.go:59]: This is fine for a first pass but let's get an issue created for this and evaluate the best course of action later.

File: release/charts/charts.go (hunk 4 above)
Comment (briandowns) [reply release/charts/charts.go:10]: Let's avoid import aliases if the imported package name doesn't conflict with any other imported or existing symbol.

File: release/charts/charts.go (hunk 10 above)
Comment (briandowns) [reply release/charts/charts.go:46]: It doesn't appear as though there needs to be additional context to these errors. What's returned from the call should suffice so no need to do additional wrapping.

File: release/charts/charts.go [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+	branchArg = "--branch-version=" + branch
+	if chart != "" {
+		chartArg = "--chart=" + chart
+	}
+
+	if err := executeChartsBuildScripts(c.Workspace, "lifecycle-status", branchArg, chartArg); err != nil {
+		return err
+	}
+
+	fmt.Printf("generated log files for inspection at: \n%s\n", c.Workspace+"/logs/")
```
Comment (briandowns) [inline release/charts/charts.go:26]: Let's remove this in favor of printing output from the calling code of this function. Library code shouldn't be putting anything to a log or STDOUT unless in debug.

**Review by briandowns — CHANGES_REQUESTED** [review]

**Comment by nicholasSUSE** [comment]
@briandowns all fixes implemented.

File: release/charts/charts.go (hunk 11 above)
Comment (nicholasSUSE) [reply release/charts/charts.go:51]: I need to change the directory back first. 

And then I need to return the output so I can print it in another function.

File: release/charts/charts.go (hunk 12 above)
Comment (briandowns) [reply release/charts/charts.go:23]: Not sure if this isn't updating but it doesn't appear as though this has been completed.

File: release/charts/charts.go (hunk 13 above)
Comment (nicholasSUSE) [reply release/charts/charts.go]: @briandowns I removed the print from here to the parent calling function, please check if this is right now.

File: cmd/release/cmd/list.go [hunk 14]
```diff
@@ -40,8 +41,40 @@ var rancherListRCDepsSubCmd = &cobra.Command{
 	},
 }
 
+var chartsListSubCmd = &cobra.Command{
+	Use:   "charts [branch] [charts](optional)",
+	Short: "List Charts assets versions state for release process",
+	RunE: func(cmd *cobra.Command, args []string) error {
+
```
Comment (briandowns) [inline cmd/release/cmd/list.go:48]: Small thing but remove this empty line.

File: release/charts/charts.go [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+	output, err := runChartsBuildScripts(c.Workspace, "lifecycle-status", branchArg, chartArg)
+	if err != nil {
+		return "", err
+	}
+
+	response := string(output) + fmt.Sprintf("\ngenerated log files for inspection at: \n%s\n", c.Workspace+"/logs/")
+	return response, nil
+}
+
+func runChartsBuildScripts(chartsRepoPath string, args ...string) ([]byte, error) {
```
Comment (briandowns) [inline release/charts/charts.go:31]: I thought I had left a comment about changing the name but I guess not. Please change this to be runChartsBuild

**Review by briandowns — CHANGES_REQUESTED** [review]

**Review by rafaelbreno — APPROVED** [review]

File: release/charts/charts.go [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+
+	response := string(output) + fmt.Sprintf("\ngenerated log files for inspection at: \n%s\n", c.Workspace+"/logs/")
+	return response, nil
+}
+
+func runChartsBuild(chartsRepoPath string, args ...string) ([]byte, error) {
+	// save current working dir
+	ecmWorkDir, err := os.Getwd()
+	if err != nil {
+		return []byte{}, err
```
Comment (briandowns) [inline release/charts/charts.go:35]: I think we can/should return nil here rather than having an unnecessary allocation.

**Review by briandowns — CHANGES_REQUESTED** [review]

**Review by briandowns — APPROVED** [review]

**Pull request merged**